### PR TITLE
Let apps register Sabre plugins or collections

### LINKED ISSUE #26195: Provide API to let apps register their own Sabre plugins
Considering that we want Webdav to be the main API in ownCloud, we need a way to let apps register their own Sabre plugins.

@DeepDiver1975 @butonic @mrow4a

### PR DESCRIPTION
**Pull request by PVince81**
## Description
See https://github.com/owncloud/core/issues/26195#issuecomment-249140878 for info.xml format.
This checks the `sabre` tags in info.xml for all apps and automatically registers the Sabre plugins and collections to the root collection.
Note that this only happens for v2 remote.php.

## Related Issue
Fixes https://github.com/owncloud/core/issues/26195

## Motivation and Context
Since we want Webdav to be used for APIs, apps need a way to register Sabre plugins and also to hook their collections to the remote.php/dav collection tree.

## How Has This Been Tested?
Manually tested with the customgroups app, not ready yet.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

@DeepDiver1975 check this out

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] TODO: also register sabre plugins in v1 endpoint

**Comment by PVince81** [comment]
- maybe let info.xml specify for which endpoint v1, v2 or both through attributes ?

**Comment by PVince81** [comment]
- [ ] TODO: attr to tell whether the plugin needs to be registered also for public webdav ? (mostly for file properties)

**Comment by PVince81** [comment]
To demo and test this feature, I tried moving the DAV plugins for the comments into the "comments" app in 64b66d78c2ae00e717a280ed4d4b8bd2ba5b4ff2

So far only the root collection seems to work. Might need further tweaking.

**Comment by PVince81** [comment]
- [ ] BUG: looks like IUserSession isn't initialized/populated yet when a plugin's `initialize()` is called, so makes it difficult to check the current user or check if user is authed (happens in CommentsPlugin):

**Comment by PVince81** [comment]
Solved the auth issue by moving the registration of app sabre plugins into `beforeMethod` handler: https://github.com/owncloud/core/pull/26761/commits/010d4a4ad56013b62a437aa5fc554c446c3c8a9e

File: apps/dav/lib/AppInfo/Application.php [hunk 1]
```diff
@@ -105,4 +122,84 @@ public function getSyncService() {
 		return $this->getContainer()->query(SyncService::class);
 	}
 
+	private function registerSabrePluginsFromApps(\OCA\DAV\Server $server) {
+		\OC_App::loadApps();
```
Comment (DeepDiver1975) [inline apps/dav/lib/AppInfo/Application.php]: load all apps? in the past we only did load apps of type filesystem to ensure some performance on the dav backend.
new app type? e.g. dav?

File: apps/dav/lib/AppInfo/Application.php [hunk 2]
```diff
@@ -105,4 +122,84 @@ public function getSyncService() {
...
+		foreach ($this->getContainer()->getServer()->getAppManager()->getInstalledApps() as $app) {
+			// FIXME: switch to public API once available
+			$appPath = \OC_App::getAppPath($app);
+			if($appPath === false) {
+				continue;
+			}
+			// FIXME: switch to public API once available
+			// load commands using info.xml
+			$info = \OC_App::getAppInfo($app);
```
Comment (DeepDiver1975) [inline apps/dav/lib/AppInfo/Application.php]: here: https://github.com/owncloud/core/pull/26588/files#diff-0513a8452fe3af4380e1bb03e0ef6aee

File: apps/dav/lib/AppInfo/Application.php [hunk 3]
```diff
@@ -105,4 +122,84 @@ public function getSyncService() {
...
+		}
+		return [];
+	}
+
+	private function loadSabrePluginsFromInfoXml($plugins) {
+		return array_map(function($plugin) {
+			try {
+				return \OC::$server->query($plugin);
+			} catch (QueryException $e) {
+				if (class_exists($command)) {
```
Comment (DeepDiver1975) [inline apps/dav/lib/AppInfo/Application.php]: $command? unknown

File: apps/dav/lib/AppInfo/Application.php [hunk 4]
```diff
@@ -105,4 +122,84 @@ public function getSyncService() {
...
+			}
+		}, $plugins);
+	}
+
+	private function loadSabreCollectionsFromInfoXml($plugins) {
+		return array_map(function($plugin) {
+			try {
+				return \OC::$server->query($plugin);
+			} catch (QueryException $e) {
+				if (class_exists($command)) {
```
Comment (DeepDiver1975) [inline apps/dav/lib/AppInfo/Application.php]: $command

**Comment by DeepDiver1975** [comment]
> TODO: also register sabre plugins in v1 endpoint

v1 only holds files - and it's just the old interface - no need to anything new there

File: apps/dav/lib/AppInfo/Application.php (hunk 3 above)
Comment (PVince81) [inline apps/dav/lib/AppInfo/Application.php]: uh oh... I copied your command code and forgot to replace

**Comment by PVince81** [comment]
@DeepDiver1975 ok then we can't port the "comments" app to be a "dav" app with Sabre plugins, because then the comments plugin would be missing in the v1 endpoint.

The other alternative is to change the web UI to use webdav v2: https://github.com/owncloud/core/pull/25494. But then if there's any client who used to expected the oc:comments-unread property on the old endpoint they'd stop working. (but the likeliness of the existence of such clients is very low)

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 ok then we can't port the "comments" app to be a "dav" app with Sabre plugins, because then the comments plugin would be missing in the v1 endpoint.

comments arn't even part of v1 as of today - right? So no need to add it

**Comment by PVince81** [comment]
> comments arn't even part of v1 as of today

Only one piece is: we have a Webdav property on files that is added through the `CommentsPlugin`. The property is there to tell how many unread comments there are to be able to display an indicator in the UI.
Without this property we'd need to do extra API calls + JS-side joins to be able to add this info to the file list.

**Comment by DeepDiver1975** [comment]
> Only one piece is: we have a Webdav property on files that is added through the CommentsPlugin. The property is there to tell how many unread comments there are to be able to display an indicator in the UI.
> Without this property we'd need to do extra API calls + JS-side joins to be able to add this info to the file list.

I see - lets hardcode this piece and we are fine

**Comment by PVince81** [comment]
Hardcode, hmmm... There are multiple levels where we can do this:
1) Keep `CommentsPlugin` within the "dav" app
or
2) Move the oc:comments-unread property logic to the `FilesPlugin`. But then it would conflict with the `CommentsPlugin` one when the "comments" app is enabled.

So I guess 1) is the best approach for now.

**Comment by PVince81** [comment]
Rebased onto master to be able to use one of the new IAppManager calls.

Remaining tasks:
- [ ] write unit tests
- [ ] move `CommentsPlugin` back to apps/dav for backward compatibility
- [ ] clear out public webdav case (could be done separately as we currently seldom register plugins for that one, but will need it eventually once we do federated metadata, etc)

**Comment by PVince81** [comment]
I refactored the whole thing to simplify it.
Wrote a unit test for the plugin manager.

Please review @DeepDiver1975 @jvillafanez @VicDeo @PhilippSchaffrath 

First have a look at https://github.com/owncloud/core/pull/26761/commits/07d43a539477c8c767090fbc10e9f94220b4b87a which brings the Sabre plugins registration code.
Then see how it's used in https://github.com/owncloud/core/pull/26761/commits/ace036b9241b851b2066205b9e60e3287e98ffe5 where the comments app is changed to use it.

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**